Fixes #1225 - Created a single ZooNode that tracks changes for all table configs

jzgithub1 opened the pull request:
@ivakegg  has reviewed this code and has recommended this pull request.  It is nearly an exact implementation of what @ctubbsii  has specified at the top of the Issue #1125. The solution has been tested in Fluo-uno and Fluo-muchos on EC-2.  It has been tested in the accumulo shell by using the 'config' command to change table properties and see the changes made effective. Other commands like 'createtable' and 'deletetable' and 'clonetable' have been tested to check that the system works as expected.  Accumulo-Testing 'cingest createtables' and 'cingest ingest' both run well.  It passes the  'mvn clean verify -Psunny' build.

The non-existence of the watchers has been verified using the ZooKeeper 'wchp' four letter word command.   These four letter commands need to be whitelisted in your zoo.cfg file by adding this line:  
 4lw.commands.whitelist=*
You need to stop and start Zookeeper to see the whitelisting commands come into effect. 
You run the 'wchp' four letter command from the shell like so:
echo wchp | nc localhost 2181 | grep 'conf/table\.
 
The only watchers remaining for table configuration are the ones  under /accumulo/{InstanceID}/namespaces.  We no longer see accumulation of table configuration item watchers as tables are created and deleted.  

This pull request should fix some system issues we have seen over the years in large systems.

keith-turner commented on core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java:
```diff
@@ -191,6 +208,177 @@ public void process(WatchedEvent event) {
+        return;
+
+      try {
+        Stat versionStat = getZooKeeper().exists(event.getPath(), watcher);
```
seems like this could be removed and the watcher passed to getData

keith-turner commented on core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java:
```diff
@@ -191,6 +208,177 @@ public void process(WatchedEvent event) {
+        }
+
+      } catch (Exception e) {
+        log.error("Error getting data from TABLE_CONFIGS zoonode " + e.getMessage());
```
Why not throw the exception?

keith-turner commented on core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java:
```diff
@@ -148,14 +159,20 @@ private ZooKeeper getZooKeeper() {
       if (log.isTraceEnabled()) {
         log.trace("{}", event);
       }
+      setWatcherForTableConfigVersion(event);
```
Will an event always trigger this and set it up?  Could this be called in constructor instead?

jzgithub1 replied:
I see what you are saying but unfortunately I don't have way to obtain the Accumulo Instance ID in the ZooCache constructor.   The zoopath of any event that occurs will have the Instance Id in it which I can 'regex out'.   I also am sure that ZooKeeper is connected and functioning by setting the watcher for the 'table-config-version' znode at this point in execution.  

I looked hard for alternatives to doing it this way.  Trust me.

keith-turner commented:
Seems like ZooCache.getChildren may still set watchers.

keith-turner commented on server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java:
```diff
@@ -629,6 +629,8 @@ private static void initZooKeeper(Opts opts, String uuid, String instanceNamePat
     zoo.putPersistentData(zkInstanceRoot, EMPTY_BYTE_ARRAY, NodeExistsPolicy.FAIL);
     zoo.putPersistentData(zkInstanceRoot + Constants.ZTABLES, Constants.ZTABLES_INITIAL_ID,
         NodeExistsPolicy.FAIL);
+    zoo.putPersistentData(zkInstanceRoot + Constants.ZTABLE_CONFIG_VERSION, new byte[0],
```
Upgrade code will also need to create node.

jzgithub1 replied:
I am working on the upgrade code part now.

jzgithub1 replied:
> Upgrade code will also need to create node.

keith-turner replied:
Just the  Upgrader9to10.upgradeZookeeper class.

keith-turner commented:
This could be follow on work and I think these would go hand in hand.  Instead of just watching a single node, we could watch 10 or 20 nodes and hash properties between them in some way.  My thinking is that this may handle high frequency changes more gracefully.  I also think it would be better to pull this code out of ZooCache into its own class.  Something like TablePropertyCache.  Having it in its own class would make it easier to make changes like the hashing thing.  Also I think it would be easier to understand ZooCache and TablePropertyCache if separate.  Again this can be follow on work.

jzgithub1 commented:
I agree with your comment above.  The high frequency changes concern me also.  When you see in the logs that the code is actually processing the changes in the proper consecutive order it is pretty amazing.  In a large network with a little latency the watch events might not inform all the tservers as perfectly as they do on a 5 node cluster in EC2.

jzgithub1 commented:
> Seems like ZooCache.getChildren may still set watchers.

My last commit deals with this in TableManager.removeTable by performing this call:
` zoo.getZooKeeper().getChildren(tableConfigPath, false);`

jzgithub1 commented:
I have another branch on my fork that solves the "Unremoved watcher problem".  It is much less risky than this one and simpler to understand.  It is called accumulo-1220.  The main way it works is with the that call to getChildren(tableConfigPath, false).  It removes all of the watches for Table config Items in the conf directory.  The other part is in the ZooCache.  @ivakegg  is looking at it now so you don't have to if you don't want to.

keith-turner commented:
Currently the watched node has a single prop name in its value.  It could have a list in its value and that could be very beneficial for create, clone, and import table. The list could help avoid missing updates.  It may also be useful for deletetable.  It would also better support a future API call that allows setting multiple props at once.

The create table operation knows of all of the props its going to set in zookeeper.  It could set all of those props (without updating the watched node for each prop) and then after setting all, it could set the value of the watched node to the list of props it updated.

jzgithub1 commented:
> Currently the watched node has a single prop name in its value. It could have a list in its value and that could be very beneficial for create, clone, and import table. The list could help avoid missing updates. It may also be useful for deletetable. It would also better support a future API call that allows setting multiple props at once.
> 
> The create table operation knows of all of the props its going to set in zookeeper. It could set all of those props (without updating the watched node for each prop) and then after setting all, it could set the value of the watched node to the list of props it updated.

That's a great idea.  Do you want me to start implementing that?  The data being stored in the table-config-version znode can't be allowed to get to large and the name should be changed to something more in line with what it is doing.

In the short term, like I mentioned, I have another branch that is lower risk than this one.

keith-turner commented:
> That's a great idea. Do you want me to start implementing that?

It could be done in this PR, if not I can open follow up issues.

jzgithub1 commented:
> > That's a great idea. Do you want me to start implementing that?
> 
> It could be done in this PR, if not I can open follow up issues.

I will start working on it now then.

jzgithub1 commented:
> Currently the watched node has a single prop name in its value. It could have a list in its value and that could be very beneficial for create, clone, and import table. The list could help avoid missing updates. It may also be useful for deletetable. It would also better support a future API call that allows setting multiple props at once.

@keith-turner ,  when I started to implement this I took another hard look at what actions actually call TablePropUtil.setTableProperty and whether it is really necessary for all of them to send notification to all the ZCacheWatchers to notify of a them of creating, cloning and importing.  The reality is that the ZCacheWatchers don't need to get all these messages on the table-version-config znode.  The ZCache never got those messages for the table configuration items until after Zcache.get was first called on the table configuraton zNodes.  

Lets look at cloning for example.  A table exists, call it sourceTable.  Its configs are in Zookeeper.  You want to clone it into destinationTable.  sourceTable has it's table configuration items in all the ZooCaches across the cluster.  TableManager.cloneTable(sourceTable, destinationTable , ..... ) gets called.  cloneTable iterates through sourceTable's config values and sets identical ones for destinationTable in Zookeeper via TablePropUtil.setTableProperty.  At this time none of the ZooCache's anywhere have these zPaths for the destinationTable inside of their internal cache objects.  destinationTable is a new table and no  Tserver or anything has ever called the function ZooCache.get for any of destinationTables zooNode data. It could not have happened.  destinationTable just came into existence.  None of destinationTable's table configs will be in any ZooCache during the execution of cloneTable.    During the cloneTable table operation if we tell the ZooCaches to remove these new configuation items from their cache it is pointless since the configuration items can not be in the ZooCache's anywhere.  That means we don't have to modify the 'table-config-version' znode during cloning because it is pointless.  It really does nothing and is not required when you think about it. Look at what the NodeDataChanged event does now regarding the 'table-config-version' node. It just deletes from the cache the table config item pulled from the data of 'table-config-version' znode.  If you have a big list of items they will just be a big list that aren't in the cache yet.     

The call paths that I have identified that require the update of the 'table-config-version' node are ones that start from MasterClientServiceHandler.removeTableProperty and MasterClientServiceHandler.setTableProperty.  They work individually.  They can't be "batched".  
I will implement "batching" when I can find a suitable entry point to do it.

Anytime you use the mechanism of changing the data in 'table-version-config' to trigger watch events in the caches to remove items that are not even there yet, you are wasting network resources and taxing the ZooKeeper server. 

@keith-turner,  you motivated me to realize a very good optimization for this process.  Thank you. 
I will check in some a new commit for you to look at.  It doesn't implement batching but it is an improvement to the last commit.

keith-turner commented:
@jzgithub1 that may be good idea to not trigger on table creation and clone. I would want to do some investigation of a few things.

Tserver execute random operations for clients.  Wondering if the following is possible?  I think there may be some things that would prevent this, but I am not sure.

  * Client process P1 initiates create table op for table TAB1
  * Client process P2 request request table props for TAB1, this dispatches to tserver TS1
  * The master process has written some props to ZK for table TAB1
  * Tserver TS1 reads a partial list of props for TAB1 into its zoocache
  * The master process completes writing props to ZK for table TAB1
  * If this could happen the partial props in ZooCache may not update 

Need to investigate the order of information put into ZK, like if the table ID to name map is put into ZK before the props that would make it visible to clients. Also do clients ignore tables in the NEW state?  I think they may, but not sure.

jzgithub1 commented:
@keith-turner I am rolling back the last commit.   We need the version of the table-config-version updated whenever a table config item  is added by creation or cloning in order to keep it updated in the 'lastTableConfigVersion' member of the ZooCache class.   The act of trying to remove items from the ZooCache that we know can't be in the ZooCache yet is not costly computationally and will be fixed later on.  The reasoning I lay out above regarding about being able to rely on ZooCache.get to actually return up-to-date config items when they are actually needed still applies.

jzgithub1 commented:
@keith-turner.  Something that I found out about cloneTable ( at least with the 'ci' table in accumulo-testing) is that when line 241 in TableManager.cloneTable gets executed there aren't any table properties to set:
```-java
log.info(
        "TableManager.cloneTable has  " + propertiesToSet.keySet().size() + " table props to set");

for (Entry<String,String> entry : propertiesToSet.entrySet())
      TablePropUtil.setTableProperty(context, tableId, entry.getKey(), entry.getValue());
```

My logs:
`install/logs/accumulo/master_dragonfly.log:2019-11-18 12:18:38,494 [tables.TableManager] INFO : TableManager.cloneTable has  0 table props to set`

jzgithub1 commented:
@keith-turner,  notifying the ZooCache of in a single notification about all of the added properties has been implemented.  As mentioned above, the only thing that really gets accomplished is that the ZooCaches find out there is a table-config-version they need to update their cache.   We are getting close to resolving issues of concern in this pull request.  

I am continuing to look into the other concerns you have bulleted above.

jzgithub1 commented:
> @jzgithub1 that may be good idea to not trigger on table creation and clone. I would want to do some investigation of a few things.
> 
> Tserver execute random operations for clients. Wondering if the following is possible? I think there may be some things that would prevent this, but I am not sure.
> 
>     * Client process P1 initiates create table op for table TAB1
> 
>     * Client process P2 request request table props for TAB1, this dispatches to tserver TS1
> 
>     * The master process has written some props to ZK for table TAB1
> 
>     * Tserver TS1 reads a partial list of props for TAB1 into its zoocache
> 
>     * The master process completes writing props to ZK for table TAB1
> 
>     * If this could happen the partial props in ZooCache may not update
> 
> 
> Need to investigate the order of information put into ZK, like if the table ID to name map is put into ZK before the props that would make it visible to clients. Also do clients ignore tables in the NEW state? I think they may, but not sure.

@keith-turner,  where we are now in this PR is it that we trigger updates of the cache on create table and clone table.  I implemented the sending of all the configs at cloning like you requested (with a caveat I will explain later) .  Delete table is going to modify the ZooCache the way it used to.  Just removing every item from ZooCache with a key that is prefixed with /accumulo/{instance_id}/tables/{table-id}.

The sequence of steps that you are talking about is something that you probably observed and I believe @ivakegg  has seen also.  It may actually be fixed by this pull request already.  This is because when a table configuration item is modified by a client process now it will definitely trigger an watch event in the ZCacheWatcher (for the table-config-version node) where it might not have been triggered before .   Only non-existent znodes that had a watcher put on them by a call to ZooCache.get(zPath) would trigger a NodeCreated event catchable in the ZCacheWatcher.  So what people may have been observing is that when they set table configurations in the accumulo shell (during clonetable or createtable), the defaults already in the ZooCache were never wiped and refreshed because there was no proper mechanism for it.  This is my initial hypothesis to explain the phenomenon of non durable table configuration modifications.  

I will continue to come up with actual steps and scenarios I can go through to address the concerns you have made in the bulleted comments above.

Please take a look at my last couple of commits to see that I am going in the right direction towards addressing other requests you have made.

jzgithub1 commented:
I observed that when a NodeChildrenChanged event occurred on a table configuration znodes (ones that ends in conf) the watcher was not being reset on that zpath and the children of the conf zpath where not being removed from the cache.  I now perform those actions.  It may fix some issues people have see with config changes not being applied because old or default configs have not been removed from the cache.

jzgithub1 commented:
> @jzgithub1 that may be good idea to not trigger on table creation and clone. I would want to do some investigation of a few things.
> 
> Tserver execute random operations for clients. Wondering if the following is possible? I think there may be some things that would prevent this, but I am not sure.
> 
>     * Client process P1 initiates create table op for table TAB1
> 
>     * Client process P2 request request table props for TAB1, this dispatches to tserver TS1
> 
>     * The master process has written some props to ZK for table TAB1
> 
>     * Tserver TS1 reads a partial list of props for TAB1 into its zoocache
> 
>     * The master process completes writing props to ZK for table TAB1
> 
>     * If this could happen the partial props in ZooCache may not update
> 
> 
> Need to investigate the order of information put into ZK, like if the table ID to name map is put into ZK before the props that would make it visible to clients. Also do clients ignore tables in the NEW state? I think they may, but not sure.

@keith-turner . my last commit assures that once the master process completes writing props to ZK for TAB1 you won't see a partial props update.  This because a NodeChildrenChanges event will occur for the specific table's 'conf' znode and all the table props will now be cleared from the cache and restored for just that table.  That didn't happen before.  All that happened before was that the value in the cache for the 'conf' directory was removed from the cache.  No other attempt was even made to put a the watcher back on that znode.  In the scenario you describe we are more solid now than we have been in the past in regards to CAP considerations. JMHO.  I have traced the logging as I created, cloned and added and modified props to tables.  I was surprised to see us handling the NodeChildrenChanged event on the 'conf' znodes the way we were.

jzgithub1 commented:
> Seems like ZooCache.getChildren may still set watchers.

From the ZooKeeper documentation:

A watch object, or function/context pair, will only be triggered once for a given notification. For example, if the same watch object is registered for an exists and a getData call for the same file and that file is then deleted, the watch object would only be invoked once with the deletion notification for the file.

[http://zookeeper.apache.org/doc/current/zookeeperProgrammers.html#sc_WatchRemoval](url)

ctubbsii commented:
I think this change should be postponed until after #1454 , which should make implementing this a bit simpler.

milleruntime commented:
This PR is outdated and is overridden by https://github.com/apache/accumulo/pull/2569
